examples/increment-only-counter/state.go

```diff
@@ -0,0 +1,91 @@
+package main
+
+import (
+	"encoding/json"
+	"sync"
+
+	"github.com/weaveworks/mesh"
+)
+
+// An increment-only counter for an arbitrary keyspace.
+type state struct {
+	mtx sync.RWMutex
+	set map[string]int
+}
+
+// state implements GossipData.
+var _ mesh.GossipData = &state{}
+
+// Construct an empty state object, ready to receive updates.
+// This is suitable to use at program start.
+// Other peers will populate us with data.
+func newState() *state {
+	return &state{
+		set: map[string]int{},
+	}
+}
+
+// Encode serializes our complete state to a slice of byte-slices.
+// In this simple example, we use a single JSON-encoded buffer.
+func (st *state) Encode() [][]byte {
+	st.mtx.RLock()
+	defer st.mtx.RUnlock()
+	buf, err := json.Marshal(st.set)
+	if err != nil {
+		panic(err)
+	}
+	return [][]byte{buf}
+}
+
+// Merge merges the other GossipData into this one,
+// and returns our resulting, complete state.
+func (st *state) Merge(other mesh.GossipData) (complete mesh.GossipData) {
+	return st.completeMerge(other.(*state).copy().set)
```

Comment:
This is copying the wrong thing. It's the LHS that needs copying. See, e.g. https://github.com/weaveworks/weave/blob/master/nameserver/entry.go#L251.